### ISSUE TITLE
Installation fails on Linode due to broken Linode configuration profile

### ISSUE THREAD

**Issue by vrms** [jlund/streisand#500]
I am a novice with Streisand and exited to get it running. I hope I am not just missing anything obvious.

Trying to spin up Streisand on Linode. My local machine is a clean xubuntu 16.04 VM, prepared as per the Streisand README 

The installation fails at the same point towards the end (executing `~/playbooks/roles/wireguard/tasks/main.yml` as it seems) for several attempts in a row now

The error I am geting from the streisand install script

`TASK [wireguard : Enable the WireGuard service so it starts at boot, and bring up the WireGuard network interface] ***
fatal: [139.162.10.186]: FAILED! => {"changed": false, "failed": true, "msg": "Unable to start service wg-quick@wg0-server.service: Job for wg-quick@wg0-server.service failed because the control process exited with error code. See \"systemctl status wg-quick@wg0-server.service\" and \"journalctl -xe\" for details.\n"}`

on my linode machine's Terminal

`root@ubuntu:/# systemctl status wg-quick@wg0-server.service
● wg-quick@wg0-server.service
   Loaded: not-found (Reason: No such file or directory)
   Active: inactive (dead)`

pretty stuck with this but saw that the wireguard website issues a Warning which basically says not to "... rely on WireGuard at this stage"

did I just hit a bad moment where wireguard it's just not working? or is there any cure for this?

**Comment by jlund** [comment]
Is this on a Xenial (16.04) server?

**Comment by vrms** [comment]
on the linode remote server yes
`root@ubuntu:~# cat /etc/os-release 
NAME="Ubuntu"
VERSION="16.04.1 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.1 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
UBUNTU_CODENAME=xenial'

my local machine however is an xubuntu 16.04 (xfce4 Desktop)

**Comment by vrms** [comment]
I have some friends @albertogalan & @rosscournoyer602  who will basically atempt the same thing today ... curiouos whether they hit the same problem

**Comment by jlund** [comment]
Cool! What region? That could also be a factor. I will try to reproduce this later tonight. There weren't any issues with Linode earlier this week when I spun up some new servers there.

As an aside, WireGuard is a new project but it is more than ready to use. It's being used quite heavily in Production at lots of companies, and for anti-censorship purposes it's a fantastic option. As with anything new though, there will inevitably be issues. We will get this one worked out :)

**Comment by vrms** [comment]
we are in China

**Comment by jlund** [comment]
That's great. I'd love to hear more feedback about WireGuard in China and whether or not it works (or how well).

In what Linode datacenter region were you creating your instance?

**Comment by vrms** [comment]
my linode is in Singapore. I think @albertogalan 's is in Japan

**Comment by jlund** [comment]
This is a Linode issue. Here's a verbatim transcript of the ticket I just opened:

> **New Ubuntu 16.04 servers start up with an out-of-date kernel**
>
> By default, newly launched Linode servers come up with kernel `4.8.6-x86_64-linode78` which was compiled in November of last year. Since then, numerous security updates have happened.
> 
> Additionally, these instances do not have the required Linux headers installed, nor are compatible Linux header packages available:
> 
> ```
> root@ubuntu:~# sudo apt-get install linux-headers-`uname -r`
> Reading package lists... Done
> Building dependency tree
> Reading state information... Done
> E: Unable to locate package linux-headers-4.8.6-x86_64-linode78
> E: Couldn't find any package by glob 'linux-headers-4.8.6-x86_64-linode78'
> E: Couldn't find any package by regex 'linux-headers-4.8.6-x86_64-linode78'
> ```
> 
> The combination of these two issues creates a result where newly launched servers are insecure and also not ready to handle certain workloads that require custom kernel modules.
> 
> Is there any reason that newly launched servers don't default to using the "GRUB 2" Kernel option in their Configuration Profile? That way people would be able to easily update their kernel so it wasn't over three months old, and the standard Linux headers package that comes installed by default (and is kept up-to-date by APT) would allow module compilation. Manually switching to a "GRUB 2"-based Configuration Profile solves all of the issues, but common tools like Ansible still launch new Linode servers without this option.
> 
> Thanks for your help and time!

I really hope that Linode can get it together soon. Between this terribly broken configuration on one of the most popular Linux distros and the [recent IP blacklisting issues](https://github.com/jlund/streisand/issues/481), I continue to strongly consider temporarily removing them from the list of supported providers.

**Comment by jlund** [comment]
I just committed a temporary hack that helps mitigate this issue on Linode. WireGuard will be broken until they fix their kernel/configuration, but everything else should work.

**Comment by vrms** [comment]
ok, thanks so much. 

Testing your new 'master' right now and also have opened a ticket with Linode support about the outdated kernel you mention.

Will let you know what comes out of both

**Comment by jlund** [comment]
Sure thing. Happy to help.

It's less the fact that the kernel is outdated (although that is a problem from a security perspective) and more the fact that they're not providing headers for their custom kernel. I don't even understand why they're doing the custom kernel thing to begin with. The stock Ubuntu 16.04 kernel works perfectly.

**Comment by rmcintosh** [comment]
4.9.7 is now live in the Linode Manager (4.9.0 has been for some time). 4.9.7 will be marked as "Latest" once it's been in testing for a bit of time. I can't really comment on the headers issue, as I'm not part of the team that builds the kernels, but we're generally moving towards supporting distro-supplied kernels by default.

**Comment by vrms** [comment]
ignoring the aforementioned wireguard error the installation completes now fine. Will be able to start exploring how to actually use it tomorrow)

for now .. thanks a ton @jlund

**Comment by vrms** [comment]
@rmcintosh    I understand now how you can apply a different kernel for an existing Linode (inside the Linode Manager > open the Linode in question > Edit > Slect kernel of choice in dropdown > reboot the Linode)

The problem at hand here actually is that the streisand scipt spins up a new node from  local machine. Is there any way to ensure the most recent kernel (and headers) are active for such? I couldn't find any Setting that says something like `use this and that kernel for new ubuntu (or whatever your distro of choice might be) nodes by default`

@rmcintosh  
>we're generally moving towards supporting distro-supplied kernels by default.

any idea how long this journey might take?

**Comment by rmcintosh** [comment]
@vrms Sorry for the wait! By default, Linode's use whatever kernel is marked "Latest" by us. If there's a more recent one, there's not an easy way to ensure that it would be used for deployments. I can see the advantage of a "Use the most recent kernel available even if it's not marked latest" feature, though. I'm happy to pass that idea along.

> any idea how long this journey might take?

As of ~2 days ago, all of our recommended distros (they show up under "64-Bit Distributions - Recommended" in the deployment dropdown) now support booting into GRUB2 to use the distribution's supplied kernel rather than ours. There are a few caveats (which we'll be resolving in subsequent image updates):

_**Fedora and CentOS** need to run the SELinux filesystem relabel. You can deploy CentOS 7 and boot directly with GRUB2 with no problems. But Fedora needs to run the relabel process, so you'd first need to boot the Linode kernel so the /.autorelabel can be formed. The relabel process must be performed whenever rebooting from the Linode to default kernels._

**_Arch, CentOS, Fedora and Gentoo** have no fd0 errors. For the **remaining distributions**, these appear in the console after selecting the kernel in the Grub menu. They're benign, but delay the boot process by about 7 seconds. Once grub is updated from a kernel upgrade, or after being manually run, the fd0 errors are gone for good. For the four distributions mentioned, we make our own Grub entry with a file at /etc/grub.d/09_default. During build, grub is updated with that script marked as executable, then chmod -x so it's not brought into later updates. The file is left in the image so customers have it as a reference if they poke around in grub.cfg.
 
When Grub is updated, either naturally or manually, our boot entry is gone and you're left with just the available kernels in /boot and their entries in Grub. The goal going forward is to eliminate these entirely from the beginning._

**Comment by vrms** [comment]
thanks for the feedback @rmcintosh 

> Linode's use whatever kernel is marked "Latest" by us. If there's a more recent one, there's not an easy way to ensure that it would be used for deployments. I can see the advantage of a "Use the most recent kernel available even if it's not marked latest" feature, though. I'm happy to pass that idea along.

hm, whether your label 'latest' is put on an (even in your world) outdated kernel  it's either a wrong label or wrong labeling practice. I think this is on purpose so I guess you'd be clearer by using a 'default' label for example instead of 'latest'. The it's at least clear that Linode has an outdated kernel as it's default

**Comment by zx2c4** [comment]
The issue is probably that Linode doesn't provide kernel headers. I've got a workaround for this. I'll try to make a script to get the headers prepared in the coming days.

**Comment by vrms** [comment]
@zx2c4 yes, that's also what @jlund supected. @rmcintosh will there be a cure for the missing headers from Linode's end as well any time?

**Comment by rmcintosh** [comment]
If you need kernel headers then you should use the distribution's kernel rather than the one supplied by the hypervisor. Simply booting into GRUB2 should accomplish that if you're running one of the "recommended" images.

**Comment by zx2c4** [comment]
Hi @rmcintosh -- didn't realize you're the Linode guy. There are two things you can do that will improve the state of affairs for everybody:

1. Enable `/proc/config.gz`. I believe Linode kernels already do this, but please double check this and ensure it's happening.
2. Make `Module.symvers` from your kernel builds available for download by users, or include it somehow with the booted system.

If you do these two things, users will be able to download the kernel source and build out of tree modules. Your competitor Scaleway does exactly this, and [even provides instructions](https://github.com/scaleway/kernel-tools#how-to-build-a-custom-kernel-module) on how to use `/proc/config.gz` together with `Module.symvers` to download kernel source and make an environment suitable for module building. For this reason, some WireGuard project infrastructure is hosted on Scaleway, not on Linode; we like to work with companies who make development simple.

Please chime back in on this thread once you've ensured that `/proc/config.gz` is enabled and made `Module.symvers` available for each kernel you build. These are both extremely easy things to do, and shouldn't take more than a few minutes of twiddling. Until this is done, Linode will not be a friendly place for interesting kernel developments.

**Comment by rmcintosh** [comment]
> Enable /proc/config.gz. I believe Linode kernels already do this, but please double check this and > ensure it's happening.

This has long been enabled -- I don't need to double check.

> Make Module.symvers from your kernel builds available for download by users, or include it 
> somehow with the booted system.

I can push for this, though, my influence/involvement in the guest kernel build process is pretty limited.

**Comment by zx2c4** [comment]
> I can push for this, though my influence/involvement in the guest kernel build process is pretty limited.

Cool, thanks.

**Comment by vrms** [comment]
> If you need kernel headers then you should use the distribution's kernel rather than the one supplied by the hypervisor. Simply booting into GRUB2 should accomplish that if you're running one of the "recommended" images.
possible to add this routine to the installation script? Unfortunately I miss the chops to attempt such myself

**Comment by vrms** [comment]
If you need kernel headers then you should use the distribution's kernel rather than the one supplied by the hypervisor. Simply booting into GRUB2 should accomplish that if you're running one of the "recommended" images.

possible to add such a routine to the installation script @jlund ? 
Due to lack of chops I can't attempt the myself though unfortunately

**Comment by kylebrowning** [comment]
This is still an issue :/

**Comment by zx2c4** [comment]
Solution: Change the Linode boot mode from one of their kernels to "GRUB2":

![image](https://cloud.githubusercontent.com/assets/10643/24467165/b5a944ec-14b3-11e7-9239-f16054ce076a.png)

Then you can simply use the distro kernel and everything works like usual.

**Comment by kylebrowning** [comment]
@zx2c4 and then re-provission?

**Comment by zx2c4** [comment]
You can change this after provisioning.

**Comment by zx2c4** [comment]
Alright I traced how to fix this in Streisand.

Streisand uses Ansiable.
Ansiable has a core module called `linode`.
The ansiable `linode` module uses `py-linode-api`.
`py-linode-api` supports setting the kernel configuration, as exposed in `linode/objects/linode/config.py`.

So, wire up ansiable's linode module to this particular key. Then set the kernel key to "grub2" in streisand. Then it will work.

If somebody cares to do this, the first step is a pull request against ansiable-core. The second is a pull request against this project. If you do this, please do take care to cross reference issues, so we're notified here.

**Comment by sjug** [comment]
Are you spelling ansible wrong on purpose?

**Comment by rmcintosh** [comment]
@zx2c4 ansible's linode module already supports the ability to specify a kernel ID during deployment, no? maybe I'm misunderstanding what you mean

**Comment by sjug** [comment]
@rmcintosh The [Linode ansible module](https://docs.ansible.com/ansible/linode_module.html) does not support kernel ID.

**Comment by rmcintosh** [comment]
@sjug yup - you're right, my mistake

**Comment by sjug** [comment]
@rmcintosh On the plus side they do have some code for creating a config (which is the necessary object to specify kernel). https://github.com/ansible/ansible/blob/eb1214baad0cbe2c4b7304caebb9ae1c7dc0d8db/lib/ansible/modules/cloud/linode/linode.py#L479

**Comment by sjug** [comment]
The very strange thing about the `GRUB 2` menu entry is it doesn't get returned by avail.kernels(). It seems the `KernelID=210`

**Comment by rmcintosh** [comment]
@sjug when querying the API you have to specify isKVM=yes&isXen=no for GRUB2's ID to appear

**Comment by sjug** [comment]
@For your PR it seems their pattern is to validate the validity of user provided IDs, as with distribution.

**Comment by blizzbhaller** [comment]
Still not working despite using GRUB2 kernel. Getting this message when installing wireguard-dkms wireguard-tools....

First Installation: checking all kernels...
Building only for 4.4.0-36-generic
Module build for the currently running kernel was skipped since the
kernel source for this kernel does not seem to be installed.

when running ip link add dev wg0 type wireguard still returns....
RTNETLINK answers: Operation not supported

**Comment by zx2c4** [comment]
Gotta install the right kernel headers now.

**Comment by blizzbhaller** [comment]
Finally got Wireguard to work in Linode:
Did a fresh linode install, deployed the Ubuntu 16.04 image, changed the kernel to GRUB2 before booting the VPS, logged into the VPS, updated the apt repositories and upgraded, added the linux headers using the command sudo apt-get install linux-headers-`uname -r`, added the wireguard ppa (you might have to run the command sudo apt-get install software-properties-common) and installed the wireguard-dkms wireguard-tools just to save time and see if i can a wireguard interface, which I was able to do and then I finally ran the streisand playbook.

Everything's working so far and enjoying wireguard.  My only concern is will using an old kernel "4.4.0-36-generic" be a security concern down the road.  Will updating the kernel to the latest one (not the custom linode kernel) disable wireguard? (since it lives in the kernel)

**Comment by rmcintosh** [comment]
ansible change is merged, PR in for the streisand change to utilize GRUB2 kernel option by default

**Comment by zx2c4** [comment]
Upset Streisand users are now coming into #wireguard and complaining about things not working and confusing others in the room. When it's eventually found out that it's a Streisand-related issue with Linode, the reaction is, "oh, ugh."

But from reading above in this ticket, it looks like everything is all set on the Streisand side, and we're just waiting on some upstream ansible change?

**Comment by cpu** [comment]
@zx2c4 I believe #574 is blocked on https://github.com/ansible/ansible/pull/23100 being shipped in a release. So far it has only landed in devel. My own understanding of the Ansible release model was patched together ~2m ago but it seems like we missed the `v2.3.0.0-1` release from 3 days ago. The [release schedule](http://docs.ansible.com/ansible/dev_guide/developing_releases.html#release-schedule) docs say that:
> When a fix orfeature gets added to the devel branch it will be part of the next release
So I'd hope it would be in the next release.

I wonder if its worth adding a prompt or something for Linode in the Wireguard role in the interim pointing to this issue. Apologies for the additional support volume :-(

**Comment by rmcintosh** [comment]
The Ansible change will be in the 2.4 release... I think a prompt to let
users know the deal might be a good idea in the interim... happy to commit
that if it's something people are interested in

**Comment by zx2c4** [comment]
Oh dang, it's not in the 2.3 release? That's a pain. 

I guess in the meantime you could always monkey patch `api.avail_kernels()` to always return `[ { 'LABEL': 'Latest 64', 'KERNELID': 210 } ]` or something ugly like that...

**Comment by cpu** [comment]
@zx2c4 @rmcintosh Thanks for all the info! I've pulled the short-term fix discussion into a new issue; https://github.com/jlund/streisand/issues/609

**Comment by Dave-Cox** [comment]
Having this same issue with Scaleway on their version of 16.04

root@UndeadW0rm:~# sudo apt-get install linux-headers-`uname -r`
Reading package lists... Done
Building dependency tree
Reading state information... Done
E: Unable to locate package linux-headers-4.9.20-std-1
E: Couldn't find any package by glob 'linux-headers-4.9.20-std-1'
E: Couldn't find any package by regex 'linux-headers-4.9.20-std-1'

**Comment by Dave-Cox** [comment]
@zx2c4 
Yes, I realize its not supported directly, but any 16.04 should work I thought. I tried the above fix and it still errored out with the following.

`failed: [161.142.113.5] (item=[u'linux-headers-4.9.20-std-1', u'linux-headers-generic', u'wireguard-dkms', u'wireguard-tools']) => {"failed": true, "item": ["linux-headers-4.9.20-std-1", "linux-headers-generic", "wireguard-dkms", "wireguard-tools"], "msg": "No package matching 'linux-headers-4.9.20-std-1' is available"}
`

**Comment by zx2c4** [comment]
Oh, yes, Streisand will error out because it wants to install those headers which don't exist as a package. But WireGuard will work fine. `apt install wireguard-dkms wireguard-tools`.

If you want built-in Scaleway support, open a new ticket.

**Comment by Dave-Cox** [comment]
@zx2c4 Thanks, I will open a ticket but I am not sure its worth it. They dont have a lot of locations. I just needed something in France to test from. Once I install wireguard manually do I have to do anything special to get Streisand to finish and create the html docs or can I just use --limit @/home/dogbone/streisand/playbooks/streisand.retry

**Comment by zx2c4** [comment]
Just remove the two linux-headers- lines from the list of packages Streisand installs.

```diff
diff --git a/playbooks/roles/wireguard/tasks/main.yml b/playbooks/roles/wireguard/tasks/main.yml
index 24f33b0..f9c82a1 100644
--- a/playbooks/roles/wireguard/tasks/main.yml
+++ b/playbooks/roles/wireguard/tasks/main.yml
@@ -11,8 +11,6 @@
   apt:
     name: "{{ item }}"
   with_items:
-    - linux-headers-{{ kernel_release.stdout }}
-    - linux-headers-generic
     - wireguard-dkms
     - wireguard-tools
 
```

Then install Streisand normally onto the box with the prepared kernel sources.

This is the last message I'll send on the issue here. If you need help with WireGuard-specific things in general, *and not Streisand-related issues*, you can come to [our mailing list](https://lists.zx2c4.com/mailman/listinfo/wireguard). Otherwise, I'm sure plenty of others here will be willing to help you with Streisand.

**Comment by cpu** [comment]
@zx2c4 Thanks for providing some pointers.

@Dave-Cox Echoing the request to move this conversation. Since your problem isn't Linode related, and the Linode issue is solved this isn't the appropriate thread. 

I'm going to lock this thread for further conversation. If there are follow up issues to address please open new issues with a descriptive title.

Thanks all
